packages/zoe/src/evalContractCode.js

```diff
@@ -1,77 +1,74 @@
 import produceIssuer from '@agoric/ertp';
-import { assert, details } from '@agoric/assert';
+import { assert } from '@agoric/assert';
 import { producePromise } from '@agoric/produce-promise';
 import { sameStructure } from '@agoric/same-structure';
+import { importBundle } from '@agoric/import-bundle';
+import { HandledPromise } from '@agoric/eventual-send';
 
-const evaluateStringToFn = (functionSrcString, endowments) => {
-  assert.typeof(functionSrcString, 'string');
-  const nestedEvaluate = src =>
-    evaluateProgram(src, { ...endowments, nestedEvaluate });
-  const fn = nestedEvaluate(`(${functionSrcString})`);
-  assert.typeof(
-    fn,
-    'function',
-    details`"${functionSrcString}" must be a string for a function, but produced ${typeof fn}`,
-  );
-  return fn;
-};
-
-const evalContractCode = (code, additionalEndowments) => {
+const evalContractBundle = (bundle, additionalEndowments, vatPowers) => {
   // Make the console more verbose.
   const louderConsole = {
     ...console,
     log: console.info,
   };
+  function myRequire(what) {
+    if (what === '@agoric/harden') {
+      return harden;
+    }
+    throw Error(`require(${what}) not implemented`);
+  }
+  harden(myRequire);
+
+  // TODO: this should really only be console and HandledPromise
   const defaultEndowments = {
+    assert,
     console: louderConsole,
     harden,
+    Nat,
+    produceIssuer,
     producePromise,
-    assert,
     sameStructure,
-    produceIssuer,
-    Nat,
+    HandledPromise,
   };
-  const fullEndowments = Object.create(null, {
-    ...Object.getOwnPropertyDescriptors(defaultEndowments),
-    ...Object.getOwnPropertyDescriptors(additionalEndowments),
-  });
 
-  // Refill our meter each crank.
-  const { meter, refillFacet } = makeMeter();
-  const doRefill = () => {
-    // Refill the meter if it wasn't exhausted.
+  const transforms = [];
+  const meteringEndowments = {};
+  if (vatPowers && vatPowers.transformMetering && vatPowers.makeGetMeter) {
+    const { makeGetMeter, transformMetering } = vatPowers;
     // This implements fail-stop, since a contract that exhausts the meter
     // will not run again.
-    if (!meter.isExhausted()) {
-      refillFacet.combined();
-    }
-  };
+    const { getMeter } = makeGetMeter({ refillIfExhausted: false });
+    transforms.push(src => transformMetering(src, getMeter));
+    meteringEndowments.getMeter = getMeter;
+  }
+  // note: Old dapps used `makeZoe({ require })`, not `makeZoe()` or
+  // `makeZoe({}, vatPowers)`. The first argument of makeZoe appears to
+  // evalContractBundle as 'additionalEndowments'. To remain compatible with
+  // them (for a brief while), we apply 'require' last, to ignore the one
+  // that appears in additionalEndowments. The fastest/cheapest way to do
+  // this is to jam it onto meteringEndowments, even though it's not related
+  // to metering at all.
+  meteringEndowments.require = myRequire;
```

Comment:
That's the fastest/cleanest way, for sure. How much worse would it be to add `extraEndowments` to the end of the list, and put just `require` there?

Reply:
I started to do that, by adding `require: myRequire` to the end of the list, but then I realized that it's using descriptors, not simple properties, and I couldn't be bothered to look up the descriptor syntax (it's probably something like `{name : 'require', value: myRequire}`). I'll give that a try.